File: typescript/src/manifest.ts
import { z } from 'zod';

export const KindSchema = z.enum(['blocks', 'txs', 'events']);
/** Daily parquet category — one of `'blocks' | 'txs' | 'events'`. */
export type Kind = z.infer<typeof KindSchema>;

const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;
const MD5_RE = /^[a-f0-9]{32}$/;
const FILE_KEY_RE = /^daily\/\d{4}-\d{2}-\d{2}\.(blocks|txs|events)\.parquet$/;
const LOOKUP_KEY_RE = /^lookups\/[A-Za-z0-9_-]+\.(json|json\.gz|parquet)$/;

const dateString = z.string().regex(DATE_RE, 'expected YYYY-MM-DD');
const md5String = z.string().regex(MD5_RE, 'expected 32 lowercase hex chars');

/** Manifest entry for one daily parquet. The `key` is relative to the archive root. */
export const FileEntrySchema = z.object({
  date: dateString,
  kind: KindSchema,
  key: z.string().regex(FILE_KEY_RE, 'unsafe FileEntry key'),
  size: z.number().int().nonnegative(),
  md5: md5String,
  schema_version: z.string(),
});
export type FileEntry = z.infer<typeof FileEntrySchema>;

/** Manifest entry for a lookup file (e.g. function selectors, event topics). */
export const LookupEntrySchema = z.object({
  key: z.string().regex(LOOKUP_KEY_RE, 'unsafe LookupEntry key'),
  size: z.number().int().nonnegative(),
  md5: md5String,
});
export type LookupEntry = z.infer<typeof LookupEntrySchema>;

/**
 * The archive's `manifest.json` shape — single source of truth for what
 * exists in the bucket. `latest_complete_date` is the latest day for
 * which all three kinds (`blocks`, `txs`, `events`) are published.
 */
export const ManifestSchema = z.object({
  chain_id: z.number().int(),
  generated_at: z.string(),
  schema_version: z.string(),
  latest_complete_date: z.string().nullable().optional(),
  columns: z.record(z.array(z.string())).default({}),
  lookups: z.record(LookupEntrySchema).default({}),
  files: z.array(FileEntrySchema).default([]),
});
export type Manifest = z.infer<typeof ManifestSchema>;

export function manifestDates(m: Manifest, kind: Kind): string[] {
  const s = new Set<string>();
  for (const f of m.files) if (f.kind === kind) s.add(f.date);
  return [...s].sort();
}

export function manifestEntry(
  m: Manifest,
  date: string,
  kind: Kind,
): FileEntry | undefined {
  return m.files.find((f) => f.date === date && f.kind === kind);
}

File: python/src/avacache/manifest.py
"""Manifest model — the single source of truth published at bucket root."""

from __future__ import annotations

import re
from datetime import date
from typing import Literal

from pydantic import BaseModel, Field, field_validator

Kind = Literal["blocks", "txs", "events"]

_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_MD5_RE = re.compile(r"^[a-f0-9]{32}$")
_FILE_KEY_RE = re.compile(
    r"^daily/\d{4}-\d{2}-\d{2}\.(blocks|txs|events)\.parquet$"
)
_LOOKUP_KEY_RE = re.compile(r"^lookups/[A-Za-z0-9_-]+\.(json|json\.gz|parquet)$")


class FileEntry(BaseModel):
    date: str  # YYYY-MM-DD
    kind: Kind
    key: str
    size: int
    md5: str
    schema_version: str

    @field_validator("date")
    @classmethod
    def _check_date(cls, v: str) -> str:
        if not _DATE_RE.match(v):
            raise ValueError(f"invalid date {v!r}: expected YYYY-MM-DD")
        return v

    @field_validator("md5")
    @classmethod
    def _check_md5(cls, v: str) -> str:
        if not _MD5_RE.match(v):
            raise ValueError(f"invalid md5 {v!r}: expected 32 lowercase hex chars")
        return v

    @field_validator("size")
    @classmethod
    def _check_size(cls, v: int) -> int:
        if v < 0:
            raise ValueError(f"invalid size {v}: must be non-negative")
        return v

    @field_validator("key")
    @classmethod
    def _check_key(cls, v: str) -> str:
        if not _FILE_KEY_RE.match(v):
            raise ValueError(
                f"unsafe FileEntry key {v!r}: expected daily/<YYYY-MM-DD>.<kind>.parquet"
            )
        return v


class LookupEntry(BaseModel):
    key: str
    size: int
    md5: str

    @field_validator("md5")
    @classmethod
    def _check_md5(cls, v: str) -> str:
        if not _MD5_RE.match(v):
            raise ValueError(f"invalid md5 {v!r}: expected 32 lowercase hex chars")
        return v

    @field_validator("size")
    @classmethod
    def _check_size(cls, v: int) -> int:
        if v < 0:
            raise ValueError(f"invalid size {v}: must be non-negative")
        return v

    @field_validator("key")
    @classmethod
    def _check_key(cls, v: str) -> str:
        if not _LOOKUP_KEY_RE.match(v):
            raise ValueError(
                f"unsafe LookupEntry key {v!r}: expected lookups/<name>.<ext>"
            )
        return v


class Manifest(BaseModel):
    chain_id: int
    generated_at: str
    schema_version: str
    latest_complete_date: str | None = None
    columns: dict[str, list[str]] = Field(default_factory=dict)
    lookups: dict[str, LookupEntry] = Field(default_factory=dict)
    files: list[FileEntry] = Field(default_factory=list)

    def dates(self, kind: Kind) -> list[date]:
        return sorted({date.fromisoformat(f.date) for f in self.files
                       if f.kind == kind})

    def entry(self, d: str | date, kind: Kind) -> FileEntry | None:
        ds = d.isoformat() if isinstance(d, date) else d
        for f in self.files:
            if f.date == ds and f.kind == kind:
                return f
        return None
